### PR TITLE
[proxy_arp] Implement proxy ARP feature

### PR DESCRIPTION
**Pull request by volodymyrsamotiy**
Signed-off-by: Volodymyr Samotiy <volodymyrs@mellanox.com>

**What I did**
Implemented ARP proxy feature.

**Why I did it**
Feature request according to HLD https://github.com/Azure/SONiC/blob/master/doc/arp/Proxy%20Arp.md.

**How I verified it**
- Verified that it is possible to enable and disable proxy ARP on non VNET interfaces
  - Verified kernel configuration via ```proc``` FS
  - Verified VLAN flood SAI configuration for VLAN RIF
- Verified that proxy ARP is enabled by default on VNET interfaces
- Verified that ARP packets are propagated correctly when proxy ARP is enabled 

**Details if related**
N/A

### CODE REVIEW

**Comment by volodymyrsamotiy** [comment]
retest this please

**Comment by prsunny** [comment]
can you please fix lgtm alerts?

**Comment by volodymyrsamotiy** [comment]
retest this please

**Comment by volodymyrsamotiy** [comment]
retest this please

File: cfgmgr/intfmgr.cpp [hunk 1]
```diff
@@ -420,7 +454,22 @@ bool IntfMgr::doIntfGeneralTask(const vector<string>& keys,
             FieldValueTuple fvTuple("mac_addr", MacAddress().to_string());
             data.push_back(fvTuple);
         }
-        
+
+        if (!proxy_arp.empty() || !vrf_name.compare(0, strlen(VNET_PREFIX), VNET_PREFIX))
```
Comment (prsunny) [inline cfgmgr/intfmgr.cpp:458]: A small request for change : After some internal discussion, we don't want to have proxy_arp enabled explicitly for Vnets. Let that also be triggered by config_db entry. I'll make necessary modifications to requirement doc. Please remove this check on Vnets

File: cfgmgr/intfmgr.cpp [hunk 2]
```diff
@@ -272,6 +272,34 @@ void IntfMgr::removeSubIntfState(const string &alias)
...
+    {
+        SWSS_LOG_ERROR("Proxy ARP state is invalid: \"%s\"", proxy_arp.c_str());
+        return false;
+    }
+
+    cmd << ECHO_CMD << " " << proxy_arp_pvlan << " > /proc/sys/net/ipv4/conf/" << alias << "/proxy_arp_pvlan";
+    EXEC_WITH_ERROR_THROW(cmd.str(), res);
+
+    SWSS_LOG_INFO("Proxy ARP set to \"%s\" on interface \"%s\"", proxy_arp.c_str(), alias.c_str());
+
```
Comment (prsunny) [inline cfgmgr/intfmgr.cpp:299]: Suggest to remove extra line to avoid file getting bigger

File: orchagent/intfsorch.cpp [hunk 3]
```diff
@@ -226,6 +227,81 @@ bool IntfsOrch::setRouterIntfsAdminStatus(const Port &port)
...
+
+    if (port.m_type == Port::VLAN)
+    {
+        if (proxy_arp == "enabled")
+        {
+            port.m_vlan_info.vlan_flood_type = SAI_VLAN_FLOOD_CONTROL_TYPE_NONE;
+        }
+        else
+        {
+            port.m_vlan_info.vlan_flood_type = SAI_VLAN_FLOOD_CONTROL_TYPE_ALL;
```
Comment (prsunny) [inline orchagent/intfsorch.cpp:291]: I don't see this updated back to port map. Please refer [this ](https://github.com/Azure/sonic-swss/blob/master/orchagent/portsorch.cpp#L2021) after updating a field of port. In this case, i dont think you need to store this value. Just pass the `flood_type` to `setIntfVlanFloodType`

File: orchagent/intfsorch.cpp [hunk 4]
```diff
@@ -226,6 +227,81 @@ bool IntfsOrch::setRouterIntfsAdminStatus(const Port &port)
...
+        }
+
+        if (!setIntfVlanFloodType(port))
+        {
+            return false;
+        }
+    }
+
+    m_syncdIntfses[alias].proxy_arp = (proxy_arp == "enabled") ? true : false;
+
```
Comment (prsunny) [inline orchagent/intfsorch.cpp:301]: Pls remove this extra line. Just a suggestion

File: orchagent/port.h [hunk 5]
```diff
@@ -32,8 +32,9 @@ typedef std::map<sai_vlan_id_t, VlanMemberEntry> vlan_members_t;
 
 struct VlanInfo
 {
-    sai_object_id_t     vlan_oid = 0;
-    sai_vlan_id_t       vlan_id = 0;
+    sai_object_id_t vlan_oid = 0;
+    sai_vlan_id_t vlan_id = 0;
+    sai_vlan_flood_control_type_t vlan_flood_type = SAI_VLAN_FLOOD_CONTROL_TYPE_ALL;
```
Comment (prsunny) [inline orchagent/port.h:37]: I suggest to remove this attribute as it is a function handled in intforch and portorch doesn't necessarily be aware of this

**Comment by volodymyrsamotiy** [comment]
fixed review comments

**Comment by abdosi** [comment]
@volodymyrsamotiy please create PR for 201911 branch. Chery-pick has conflict

@prsunny

File: orchagent/intfsorch.cpp (hunk 3 above)
Comment (shikenghua) [reply orchagent/intfsorch.cpp]: Hi @volodymyrsamotiy @prsunny, is it alright to influence the flooding behavior of all BUM packets when proxy_arp enabled? This flood_type attr seems not only applied to ARP packets.
````
typedef enum _sai_vlan_flood_control_type_t
{
    /**
     * @brief Flood on all vlan members
     *
     * When setting all to broadcast or unknown multicast flood, it also includes
     * flooding to the router. When setting all to unknown unicast flood, it does
     * not include flooding to the router
     */
    SAI_VLAN_FLOOD_CONTROL_TYPE_ALL,

    /** Disable flooding */
    SAI_VLAN_FLOOD_CONTROL_TYPE_NONE,

    /** Flood on the L2MC group */
    SAI_VLAN_FLOOD_CONTROL_TYPE_L2MC_GROUP,

    /**
     * @brief Flood on all vlan members and L2MC group
     *
     * Flood on all vlan members, without the router
     * In addition, flood on the supplied L2MC group
     */
    SAI_VLAN_FLOOD_CONTROL_TYPE_COMBINED

} sai_vlan_flood_control_type_t;
````

**Pull request merged**